### ISSUE TITLE
Declare that parameter cannot be readonly

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/ballerina-spec#547]
With the introduction of read-only intersection types, at compile-time we can disallow langlib methods that are attempting to modify an immutable value. 

```ballerina
public function main() {
    int[] & readonly arr = [1, 2];
    int i = arr.remove(0); // error!
}
```

Similar to the original suggestion, IMO it would be useful if we have an annotation across the different langlibs (including xml and table) which would indicate that the particular method would attempt updating the original value upon which the method is called.

_Originally posted by @MaryamZi in https://github.com/ballerina-platform/ballerina-spec/issues/345#issuecomment-644894165_

**Comment by jclark** [comment]
@MaryamZi Can't we do this more generally in the type system? For example, we can say that a value v belongs to a type `!readonly` if it does not have its readonly bit set. Then we can declare the first argument to `remove` to be `!readonly & Type[]`.

**Comment by MaryamZi** [comment]
Yeah, this would be clearer too, since the signature itself would imply the constraints.

I assume `!T` will only be allowed with `readonly` though (where it specifically indicates whether the read-only bit is set), and not any arbitrary type?

**Comment by jclark** [comment]
Yes, I had been thinking that `!` would be allowed only with `readonly`.

But another approach is to support a difference operator on types (corresponding to set difference). `T1 - T2` means the set difference of T1 and T2 i.e. T1 but not T2. Then you can write it as `Type[] - readonly`. This is quite a useful operator, particularly for avoiding ambiguities with union. Some regular expression flavours support this for character classes e.g. `[a-z-[aeiou]]` to match all consonants.
